dhall-nixpkgs/Main.hs

```diff
@@ -0,0 +1,828 @@
+        )
+
+parseGitHub :: Parser GitHub
+parseGitHub = do
+    name <- parseName
+
+    uri <- Options.strArgument (Options.metavar "URL")
+
+    rev <- do
+        optional
+            (Options.strOption
+                (   Options.long "rev"
+                <>  Options.help "Git revision to use"
+                )
+            )
+
+    hash <- do
+        optional
+            (Options.strOption
+                (   Options.long "hash"
+                <>  Options.help "Expected SHA256 hash"
+                )
+            )
+
+    fetchSubmodules <-
+        Options.switch
+            (   Options.long "fetch-submodules"
+            <>  Options.help "Fetch git submodules"
+            )
+
+    directory <-
+        Options.strOption
+            (   Options.long "directory"
+            <>  Options.help "Subdirectory containing the Dhall package"
+            <>  Options.value ""
+            )
+
+    file <- parseFile
+
+    source <- parseSource
+
+    return GitHub{..}
+
+parseDirectory :: Parser Directory
+parseDirectory = do
+    name <- parseName
+
+    directory <- Options.strArgument (Options.metavar "DIRECTORY")
+
+    file <- parseFile
+
+    source <- parseSource
+
+    return Directory{..}
+
+parserInfoOptions :: ParserInfo Options
+parserInfoOptions =
+    Options.info
+        (Options.helper <*> parseOptions)
+        (   Options.progDesc "Convert a Dhall project to a Nix package"
+        <>  Options.fullDesc
+        )
+
+die :: MonadIO io => Text -> io a
+die text = liftIO $ do
+    Text.IO.hPutStr System.IO.stderr text
+
+    System.Exit.exitFailure
+
+main :: IO ()
+main = do
+    GHC.IO.Encoding.setLocaleEncoding System.IO.utf8
+
+    options <- Options.execParser parserInfoOptions
+
+    case options of
+        OptionsGitHub    github    -> githubToNixpkgs    github
+        OptionsDirectory directory -> directoryToNixpkgs directory
+
+-- | Convenient utility for generating command-line options
+toListWith :: (a -> [ Text ]) -> Maybe a -> [ Text ]
+toListWith f (Just x ) = f x
+toListWith _  Nothing  = [ ]
+
+nub :: Ord a => [a] -> [a]
+nub = Foldl.fold Foldl.nub
+
+{-| This specialization of `nub` is necessary to work around a type-checking
+    loop
+-}
+nub'
+    :: Ord (f (Fix f)) => [ (Text, Maybe (Fix f)) ] -> [ (Text, Maybe (Fix f)) ]
+nub' = nub
+
+{-| The following Nix code is required reading for understanding how
+    `findExternalDependencies` needs to work:
+
+    <https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/dhall/build-dhall-package.nix>
+
+    The Nixpkgs support for Dhall essentially replaces all remote imports with
+    cache hits, but doing so implies that all remote imports must be protected
+    by an integrity check.
+
+    This function finds all remote imports that are transitive dependencies of
+    the given expression, failing if any of them are missing integrity checks.
+-}
+findExternalDependencies :: Expr Src Import -> StateT Status Shell URL
```

Comment:
What's the purpose of the `Shell` monad here? I don't quite understand how it's being used here.

Reply:
@sjakobi: The `Turtle.Shell` module explains a bit how it works:

http://hackage.haskell.org/package/turtle-1.5.19/docs/Turtle-Shell.html

... as well as the `turtle` tutorial:

http://hackage.haskell.org/package/turtle-1.5.19/docs/Turtle-Tutorial.html#g:9